services/ingestion/sealing_verifier.go

```diff
@@ -0,0 +1,334 @@
+package ingestion
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/onflow/flow-go-sdk"
+	"github.com/onflow/flow-go-sdk/access"
+	flowGo "github.com/onflow/flow-go/model/flow"
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/onflow/flow-evm-gateway/models"
+	errs "github.com/onflow/flow-evm-gateway/models/errors"
+	"github.com/onflow/flow-evm-gateway/services/requester"
+	"github.com/onflow/flow-evm-gateway/storage/pebble"
+)
+
+var _ models.Engine = &SealingVerifier{}
+
+// SealingVerifier verifies that soft finality events received over the Access polling API match the
+// actually sealed results from the event stream.
+type SealingVerifier struct {
+	*models.EngineStatus
+
+	logger zerolog.Logger
+	client *requester.CrossSporkClient
+	chain  flowGo.ChainID
+
+	startHeight            uint64
+	eventsHash             *pebble.EventsHash
+	unsealedBlocksToVerify map[uint64]flow.Identifier
+	sealedBlocksToVerify   map[uint64]flow.Identifier
+
+	mu sync.Mutex
+}
+
+// NewSealingVerifier creates a new sealing verifier.
+func NewSealingVerifier(
+	logger zerolog.Logger,
+	client *requester.CrossSporkClient,
+	chain flowGo.ChainID,
+	eventsHash *pebble.EventsHash,
+	startHeight uint64,
+) *SealingVerifier {
+	return &SealingVerifier{
+		EngineStatus:           models.NewEngineStatus(),
+		logger:                 logger.With().Str("component", "sealing_verifier").Logger(),
+		client:                 client,
+		chain:                  chain,
+		startHeight:            startHeight,
+		eventsHash:             eventsHash,
+		unsealedBlocksToVerify: make(map[uint64]flow.Identifier),
+		sealedBlocksToVerify:   make(map[uint64]flow.Identifier),
+	}
+}
+
+// Stop the engine.
+func (v *SealingVerifier) Stop() {
+	v.MarkDone()
+	<-v.Stopped()
+}
+
+// SetStartHeight sets the start height for the sealing verifier.
+// This is used to update the height when backfilling already sea
+func (v *SealingVerifier) SetStartHeight(height uint64) {
+	v.startHeight = height
+}
+
+// AddBlock adds a block to the sealing verifier for verification when the sealed data is received.
+func (v *SealingVerifier) AddBlock(events flow.BlockEvents) error {
```

Comment:
Better to rename to `AddFinalizedBlock` because otherwise the events won't be unique by height.

Reply:
the gateway doesn't really have a notion of blocks before they are finalized. How about `AddUnsealedBlock`?

Reply:
How does gateway requests blocks from AN? Does it request only finalized blocks or any executed blocks? If later, then it's possible that there will be multiple events for the same block. However, we assumed block events is unique by height, isn't it?